File: src/types/node.ts
import type { Parameter } from './parameter'

export type NodeCategory =
  | 'sop' | 'dop' | 'cop' | 'vop' | 'lop' | 'rop' | 'chop' | 'top' | 'object'
  | 'color' | 'filter' | 'merge' | 'transform' | 'channel' | 'draw' | 'deep'
  | 'geometry' | 'shader' | 'compositor'
  | 'blueprint' | 'material' | 'animation'
  | 'utility' | 'math' | 'logic' | 'custom' | 'other'

export type NodeObjectType =
  | 'software_node'
  | 'recipe'
  | 'blueprint'
  | 'documentation'
  | 'asset'
  | 'project'
  | 'template'
  | 'reference'
  | 'learning_topic'
  | 'external_link'
  | 'note'
  | 'custom'

export interface NodePort {
  id: string
  name: string
  dataType: string
  required: boolean
  multi: boolean
  description?: string
}

export interface CortexNode {
  id: string
  vaultId?: string
  softwareId?: string
  name: string
  displayName: string
  category: NodeCategory
  objectType: NodeObjectType
  description?: string
  version?: string
  color?: string
  icon?: string
  tags: string[]
  inputs: NodePort[]
  outputs: NodePort[]
  parameters: Parameter[]
  documentation?: string
  notes?: string
  productionTips: string[]
  mediaIds: string[]
  isDeprecated: boolean
  deprecatedBy?: string
  createdAt: string
  updatedAt: string
  metadata: Record<string, unknown>
}

export interface Software {
  id: string
  name: string
  displayName: string
  version?: string
  icon?: string
  color?: string
  description?: string
  website?: string
  createdAt: string
}

export interface CreateNodeInput {
  vaultId?: string
  softwareId?: string
  name: string
  displayName: string
  category: NodeCategory
  objectType: NodeObjectType
  description?: string
  version?: string
  color?: string
  icon?: string
  tags?: string[]
  inputs?: NodePort[]
  outputs?: NodePort[]
  parameters?: Parameter[]
  documentation?: string
  notes?: string
  productionTips?: string[]
  metadata?: Record<string, unknown>
}

export interface UpdateNodeInput {
  id: string
  name?: string
  displayName?: string
  description?: string
  color?: string
  icon?: string
  tags?: string[]
  parameters?: Parameter[]
  documentation?: string
  notes?: string
  productionTips?: string[]
  metadata?: Record<string, unknown>
  category?: NodeCategory
  objectType?: NodeObjectType
  inputs?: NodePort[]
  outputs?: NodePort[]
}

File: src-tauri/src/domain/node.rs
#![allow(dead_code)]
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use super::parameter::Parameter;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Node {
    pub id: Uuid,
    pub vault_id: Option<Uuid>,
    pub software_id: Option<Uuid>,
    pub name: String,
    pub display_name: String,
    pub category: NodeCategory,
    pub object_type: NodeObjectType,
    pub description: Option<String>,
    pub version: Option<String>,
    pub color: Option<String>,
    pub icon: Option<String>,
    pub tags: Vec<String>,
    pub inputs: Vec<NodePort>,
    pub outputs: Vec<NodePort>,
    pub parameters: Vec<Parameter>,
    pub documentation: Option<String>,
    pub notes: Option<String>,
    pub production_tips: Vec<String>,
    pub media_ids: Vec<Uuid>,
    pub is_deprecated: bool,
    pub deprecated_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub metadata: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum NodeCategory {
    // Houdini categories
    Sop,
    Dop,
    Cop,
    Vop,
    Lop,
    Rop,
    Chop,
    Top,
    Object,
    // Nuke
    Color,
    Filter,
    Merge,
    Transform,
    Channel,
    Draw,
    Deep,
    // Blender
    Geometry,
    Shader,
    Compositor,
    // Unreal
    Blueprint,
    Material,
    Animation,
    // Generic
    Utility,
    Math,
    Logic,
    Custom,
    Other,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum NodeObjectType {
    SoftwareNode,
    Recipe,
    Blueprint,
    Documentation,
    Asset,
    Project,
    Template,
    Reference,
    LearningTopic,
    ExternalLink,
    Note,
    Custom,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NodePort {
    pub id: String,
    pub name: String,
    pub data_type: String,
    pub required: bool,
    pub multi: bool,
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateNodeInput {
    pub vault_id: Option<Uuid>,
    pub software_id: Option<Uuid>,
    pub name: String,
    pub display_name: String,
    pub category: NodeCategory,
    pub object_type: NodeObjectType,
    pub description: Option<String>,
    pub version: Option<String>,
    pub color: Option<String>,
    pub icon: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub inputs: Vec<NodePort>,
    #[serde(default)]
    pub outputs: Vec<NodePort>,
    #[serde(default)]
    pub parameters: Vec<Parameter>,
    pub documentation: Option<String>,
    pub notes: Option<String>,
    #[serde(default)]
    pub production_tips: Vec<String>,
    pub metadata: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateNodeInput {
    pub id: Uuid,
    pub name: Option<String>,
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub color: Option<String>,
    pub icon: Option<String>,
    pub tags: Option<Vec<String>>,
    pub parameters: Option<Vec<Parameter>>,
    pub documentation: Option<String>,
    pub notes: Option<String>,
    pub production_tips: Option<Vec<String>>,
    pub metadata: Option<serde_json::Value>,
    pub category: Option<String>,
    pub object_type: Option<String>,
    pub inputs: Option<Vec<NodePort>>,
    pub outputs: Option<Vec<NodePort>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Software {
    pub id: Uuid,
    pub name: String,
    pub display_name: String,
    pub version: Option<String>,
    pub icon: Option<String>,
    pub color: Option<String>,
    pub description: Option<String>,
    pub website: Option<String>,
    pub created_at: DateTime<Utc>,
}
